Cubical/Experiments/truncations.agda

```diff
@@ -0,0 +1,165 @@
+{-# OPTIONS --cubical --safe #-}
+module Cubical.Experiments.truncations where
+
+open import Cubical.Data.Nat.Base
+open import Cubical.Core.Prelude
+
+isOfHLevel : ∀ {ℓ} → ℕ → Set ℓ → Set ℓ
+isOfHLevel 0 A = isContr A
+isOfHLevel 1 A = isProp A
+isOfHLevel (suc (suc n)) A = (x y : A) → isOfHLevel (suc n) (x ≡ y)
+
+isGroupoid : Set → Set
+isGroupoid = isOfHLevel 3
+
+isTwoGroupoid : Set → Set
+isTwoGroupoid = isOfHLevel 4
+
+transport' : (p : I → Set) → (p i0) → (p i1)
+transport' p a = transp (λ i → p i) i0 a
+
+data propTrunc (A : Set) : Set where
+  pinc : A → propTrunc A
+  pIsProp : isProp (propTrunc A)
+
+propTruncElim : (A : Set) (B : (propTrunc A) → Set)
+                (bP : (x : propTrunc A) → isProp (B x))
+                (f : (x : A) → B (pinc x))
+                (x : propTrunc A) → B x
+propTruncElim A B bP f = proof
+  where
+  proof : (x : propTrunc A) → B x
+  proof (pinc x) = f x
+  proof (pIsProp x y i) = L i
+    where
+    pA = propTrunc A
+    L0 = bP x
+    C  = pIsProp x y
+    B' : {x y : pA} → (p : x ≡ y) → Set
+    B' {x} {y} p = (a : B x) → (b : B y) → PathP (λ i → B (p i)) a b 
+    B0 : B' (λ j → C i0) ≡ B' (λ j → C j)
+    B0 i = B' (λ j → C (i ∧ j))
+    L1 : B' C
+    L1 = transport B0 L0
+    L : PathP (λ i → B (pIsProp x y i)) (proof x) (proof y)
+    L = L1 (proof x) (proof y)
+
+data setTrunc (A : Set) : Set where
+  tinc : A → setTrunc A
+  tIsSet : isSet (setTrunc A)
+
+setTruncElim : (A : Set) (B : (setTrunc A) → Set)
+               (bS : (x : setTrunc A) → isSet (B x))
+               (f : (x : A) → B (tinc x))
+               (x : setTrunc A) → B x
+setTruncElim A B bS f = proof
+  where
+  proof : (x : setTrunc A) → B x
+  proof (tinc x) = f x
+  proof (tIsSet x y p q i j) = L i j
+    where
+    sA = setTrunc A
+    C = tIsSet x y p q
+    M1 : {x y : sA} (p : x ≡ y) → B x → B y → Set
+    M1 p a b = PathP (λ i → B (p i)) a b
+    M2 : {x y : sA} {p q : x ≡ y} (r : p ≡ q) {a : B x} {b : B y}
+       → (M1 p a b) → (M1 q a b) → Set
+    M2 r {a = a} {b = b} c d = PathP (λ i → M1 (r i) a b) c d
+    B' : {x y : sA} {p q : x ≡ y} (r : p ≡ q) → Set
+    B' {x} {y} {p} {q} r = {a : B x} {b : B y} (c : M1 p a b) (d : M1 q a b)
+                           → M2 r c d
+    L0 : B' (λ i j → x)
+    L0 c d = bS x _ _ c d 
+    B0 B1 : I → Set
+    B0 k = B' (λ i j → C i0 (j ∧ k))
+    B1 k = B' (λ i j → C (i ∧ k) j)
+    L1 = transp B0 i0 L0
+    L2 = transp B1 i0 L1
+    L = L2 (λ i → proof (p i)) (λ i → proof (q i))
+
+data groupoidTrunc (A : Set) : Set where
+  tinc : A → groupoidTrunc A
+  tIsGroupoid : isGroupoid (groupoidTrunc A)
+
+groupoidTruncElim : (A : Set) (B : (groupoidTrunc A) → Set)
```

Comment:
Can you please compress the proof of this as well?

Reply:
> Can you please compress the proof of this as well?

By compress you mean inline everything? Sure I can do it. 

The current way of writing was just make sure I get how it generalizes. Before it wasn't entirely clear to me how to move from `propTruncElim` to `g2TruncElim`.

Let me do it for `groupoidTruncElim` first to see whether the compressed one is enlightening or not.

Reply:
Yes, inline and if something is definitionally the same as `J` then use `J`